### PR TITLE
Use duktape gem as default JS engine on Windows-MINGW

### LINKED ISSUE #30014: TypeError on a newly created project with default javascript engine on Windows
The default javascript engine on Windows is Windows Script Host (JScript). Unfortunatelly this engine isn't able to process the default assets.

As a workaround both the duktape gem or Node.js can be used.

### Steps to reproduce

* Use Windows-10
* Install [RubyInstaller-2.4.1-2](https://rubyinstaller.org) (inkluding MSYS2 development tools)
* Then run:
   gem install rails
   rails new railstest2
   cd railstest2
   rails generate scaffold cats name:string
   rails db:migrate
   rails s
* Browse to http://localhost:3000/cats

### Expected behavior
It should show the index page for cats.

### Actual behavior
It raises the error: TypeError: Das Objekt unterstützt diese Eigenschaft oder Methode nicht.
(in English: The object doesn't support this attribute or method)

See the full stacktrace: https://gist.github.com/larskanis/0e23c2b327c8189712517fc3bc9e348b

### System configuration
**Rails version**: 5.1.2

**Ruby version**: ruby 2.4.1p111 (2017-03-22 revision 58053) [x64-mingw32]

### PR DESCRIPTION
**Pull request by larskanis**
The default javascript engine on Windows is Windows Script Host (JScript). However this engine isn't able to process the default assets.

The recommended engine on other platforms `mini_racer` doesn't support Windows. There are other alternatives like Node.js, but they require additional setup. So `duktape` is the [recommended engine](https://github.com/oneclick/rubyinstaller2/wiki/FAQ#q-are-there-recommendations-for-working-with-rails) on Windows.

Fixes #30014

### CODE REVIEW

**Comment by sgrif** [comment]
Should this be set for mswin as well?

**Comment by larskanis** [comment]
> Should this be set for mswin as well?

I think so, but I didn't test it. So this is the conservative setting.

**Comment by sgrif** [comment]
Would you mind testing it? It would be great to fix this for all the
windows cases.

**Comment by guilleiguaran** [comment]
I think we can cover all the Win cases with this:

`RUBY_PLATFORM.match(/bccwin|cygwin|mingw|mswin|wince/)`

**Comment by sgrif** [comment]
Or just `Gem.win_platform?`

**Comment by rafaelfranca** [comment]
I think we should check why the default engine can't compile the default assets before changing engines.

**Comment by sgrif** [comment]
The default engine cannot compile the default assets because it is only ES3 compatible, and the CoffeeScript compiler has required ES5 for several versions. We would need to inject a polyfill for at least `Object.create`, and probably several others.

**Comment by guilleiguaran** [comment]
@rafaelfranca I think I haven't seen any case of someone using JScript successfully since some years ago.

**Comment by larskanis** [comment]
The term "default engine" is a bit mis-leading for JScript - it is more a [last resort if no other engine is available](https://github.com/rails/execjs/blob/master/lib/execjs/runtimes.rb#L81-L91).

I can confirm, that the duktape-1.6.1.0.gem builds and runs fine on [RubyInstaller-2.4.1-2](https://rubyinstaller.org) and RubyInstaller-2.2.3 with DevKit.

I can also confirm, that the duktape-1.6.1.0.gem builds and runs on `ruby 2.5.0dev (2017-03-13 trunk 57962) [i386-mswin32_120]` . However I couldn't find any recent MS-Visual-C compiled Ruby binary, so I had to compile the ruby sources manually in order to test this out. Due to missing dependencies I wasn't able to install the full rails framework, but `p Duktape::Context.new.eval_string('2+3')` works as expected.


I also tried Cygwin: `ruby 2.3.3p222 (2016-11-21 revision 56859) [x86_64-cygwin]`, `ruby 2.3.3p222 (2016-11-21 revision 9808) [i386-cygwin]` and MSYS2: `ruby 2.3.3p222 (2016-11-21 revision 56859) [x86_64-msys]`, `ruby 2.3.0p0 (2015-12-25 revision 53290) [i386-msys]`. They all failed to compile any extensions with an endless message loop into `mkmf.log` while running `extconf.rb`, which is a [known issue](https://cygwin.com/faq/faq.html#faq.using.fixing-fork-failures) but I wasn't able to fix it locally:
```
"gcc -o conftest.exe -I/usr/include/ruby-2.3.0 -I/usr/include/ruby-2.3.0/ruby/backward -I/usr/include/ruby-2.3.0 -I.    -ggdb -O2 -pipe -Wimplicit-function-declaration   -std=c99 conftest.c  -L. -L/usr/lib -L. -fstack-protector     -lruby230  -lpthread -lgmp -ldl -lcrypt  "
      3 [main] ruby 18632 fork: child 17888 - died waiting for dll loading, errno 11
1485807 [main] ruby 18632 fork: child 4600 - died waiting for dll loading, errno 11
3148229 [main] ruby 18632 fork: child 1476 - died waiting for dll loading, errno 11
[... end so on]
```

To sum up: `RUBY_PLATFORM =~ /mingw|mswin/` seems to be a safe setting. Cygwin and MSYS2 will probably work with duktape equally. Maybe someone with a proper setup can verify this? The other Windows platforms `bccwin|djgpp|wince` are no longer supported by MRI.

**Comment by sgrif** [comment]
Thank you for the thorough investigation. I think we should proceed exactly as you've described.

**Comment by guilleiguaran** [comment]
@larskanis thank you so much for your research on this!!

Please update your PR to use `RUBY_PLATFORM =~ /mingw|mswin/` and we will proceed to merge it.

**Comment by larskanis** [comment]
OK, I updated the PR for mswin.
